alf-api/src/main/java/io/axway/alf/exception/ArgumentSerializationHelper.java

```diff
@@ -8,60 +8,63 @@
 
 final class ArgumentSerializationHelper {
     static void writeArgumentsToStream(@Nullable Consumer<Arguments> argsConsumer, ObjectOutputStream out) throws IOException {
-        List<Pair> arguments = new ArrayList<>();
+        SerializableArguments instance = new SerializableArguments();
         if (argsConsumer != null) {
-            argsConsumer.accept(new Arguments() {
-                @Override
-                public Arguments add(String key, @Nullable Object value) {
-                    arguments.add(new Pair(key, value));
-                    return this;
-                }
-            });
-        }
-        out.writeInt(arguments.size());
-        for (Pair pair : arguments) {
-            out.writeUTF(pair.getKey());
-            out.writeObject(pair.getValue());
+            argsConsumer.accept(instance);
         }
+        instance.writeExternal(out);
     }
 
     @Nullable
     static Consumer<Arguments> readArgumentsFromStream(ObjectInputStream in) throws IOException, ClassNotFoundException {
-        int argumentsLength = in.readInt();
+        SerializableArguments instance = new SerializableArguments();
+        instance.readExternal(in);
+        return instance.m_arguments.isEmpty() ? null : instance.asArgumentsConsumer();
+    }
 
-        if (argumentsLength == 0) {
-            return null;
-        }
+    public static final class SerializableArguments implements Arguments, Externalizable {
+        private final List<Map.Entry<String, Object>> m_arguments = new ArrayList<>();
```

Comment:
wouldn't you create the pair class SerializableArgument instead ?

Reply:
Sure, but I preferred `Entry` because it's the right semantic with key and value.